go/mysqlconn/replication/binlog_event.go

```diff
@@ -0,0 +1,317 @@
+	BinlogChecksumAlgCRC32 = 1
+	// BinlogChecksumAlgUndef indicates that checksums are not supported.
+	BinlogChecksumAlgUndef = 255
+)
+
+// These constants describe the event types.
+// See: http://dev.mysql.com/doc/internals/en/binlog-event-type.html
+const (
+	eUnknownEvent           = 0
+	eStartEventV3           = 1
+	eQueryEvent             = 2
+	eStopEvent              = 3
+	eRotateEvent            = 4
+	eIntvarEvent            = 5
+	eLoadEvent              = 6
+	eSlaveEvent             = 7
+	eCreateFileEvent        = 8
+	eAppendBlockEvent       = 9
+	eExecLoadEvent          = 10
+	eDeleteFileEvent        = 11
+	eNewLoadEvent           = 12
+	eRandEvent              = 13
+	eUserVarEvent           = 14
+	eFormatDescriptionEvent = 15
+	eXIDEvent               = 16
+	eBeginLoadQueryEvent    = 17
+	eExecuteLoadQueryEvent  = 18
+	eTableMapEvent          = 19
+	eWriteRowsEventV0       = 20
+	eUpdateRowsEventV0      = 21
+	eDeleteRowsEventV0      = 22
+	eWriteRowsEventV1       = 23
+	eUpdateRowsEventV1      = 24
+	eDeleteRowsEventV1      = 25
+	eIncidentEvent          = 26
+	eHeartbeatEvent         = 27
+	eIgnorableEvent         = 28
+	eRowsQueryEvent         = 29
+	eWriteRowsEventV2       = 30
+	eUpdateRowsEventV2      = 31
+	eDeleteRowsEventV2      = 32
+	eGTIDEvent              = 33
+	eAnonymousGTIDEvent     = 34
+	ePreviousGTIDsEvent     = 35
+
+	// MySQL 5.7 events
+	eTransactionContextEvent = 36
+	eViewChangeEvent         = 37
+	eXAPrepareLogEvent       = 38
+
+	// MariaDB specific values. They start at 160.
+	eMariaAnnotateRowsEvent     = 160
+	eMariaBinlogCheckpointEvent = 161
+	eMariaGTIDEvent             = 162
+	eMariaGTIDListEvent         = 163
+	eMariaStartEncryptionEvent  = 164
+)
+
+// IsZero returns true if the BinlogFormat has not been initialized.
+func (f BinlogFormat) IsZero() bool {
+	return f.FormatVersion == 0 && f.HeaderLength == 0
+}
+
+// HeaderSize returns the header size of any event type.
+func (f BinlogFormat) HeaderSize(typ byte) byte {
+	return f.HeaderSizes[typ-1]
+}
+
+// Query contains data from a QUERY_EVENT.
+type Query struct {
+	Database string
+	Charset  *binlogdatapb.Charset
+	SQL      string
+}
+
+// String pretty-prints a Query.
+func (q Query) String() string {
+	return fmt.Sprintf("{Database: %q, Charset: %v, SQL: %q}",
+		q.Database, q.Charset, q.SQL)
+}
+
+// TableMap contains data from a TABLE_MAP_EVENT.
+type TableMap struct {
+	Flags    uint16
+	Database string
+	Name     string
+	Columns  []TableMapColumn
+}
+
+// TableMapColumn describes a table column inside a TABLE_MAP_EVENT.
+type TableMapColumn struct {
+	Type      byte
+	CanBeNull bool
+}
+
+// Rows contains data from a {WRITE,UPDATE,DELETE}_ROWS_EVENT.
+type Rows struct {
+	// Flags has the flags from the event.
+	Flags uint16
+
+	// IdentifyColumns describes which columns are included to
+	// identify the row. It is a bitmap indexed by the TableMap
+	// list of columns.
+	// Set for UPDATE and DELETE.
+	IdentifyColumns Bitmap
```

Comment:
Yes, no memory copy, no allocation. We just point to the data of the original packet.